runtime/commonMain/src/kotlinx/serialization/SealedSerializer.kt

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2017-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.serialization
+
+import kotlinx.serialization.internal.AbstractPolymorphicSerializer
+import kotlinx.serialization.internal.SerialClassDescImpl
+import kotlinx.serialization.modules.SerializersModule
+import kotlin.reflect.KClass
+
+/**
+ * This class provides support for multiplatform polymorphic serialization for sealed classes.
+ *
+ * In contrary to [PolymorphicSerializer], all known subclasses with serializers must be passed
+ * in `subclasses` and `subSerializers` constructor parameters. This action is automatically performed by a compiler plugin
+ * when you mark sealed class as `@Serializable`. If a subclass is sealed class itself, all its subclasses are registered, too.
+ * In most of the cases, you won't need to perform any additional manual setup:
+ *
+ * ```
+ * @Serializable
+ * sealed class SimpleSealed {
+ *     @Serializable
+ *     public data class SubSealedA(val s: String) : SimpleSealed()
+ *
+ *     @Serializable
+ *     public data class SubSealedB(val i: Int) : SimpleSealed()
+ * }
+ *
+ * // will perform correct polymorphic serialization and deserialization:
+ * Json.stringify(SimpleSealed.serializer(), SubSealedA("foo"))
+ * ```
+ *
+ * However, it is still possible to register additional subclasses later using regular [SerializersModule].
+ * It is required when one of inheritors of sealed class is an abstract class:
+ *
+ * ```
+ * @Serializable
+ * sealed class ProtocolWithAbstractClass {
+ *     @Serializable
+ *     abstract class Message : ProtocolWithAbstractClass() {
+ *         @Serializable
+ *         data class StringMessage(val description: String, val message: String) : Message()
+ *
+ *         @Serializable
+ *         data class IntMessage(val description: String, val message: Int) : Message()
+ *     }
+ *
+ *     @Serializable
+ *     data class ErrorMessage(val error: String) : ProtocolWithAbstractClass()
+ * }
+ * ```
+ *
+ * In this case, `ErrorMessage` would be registered automatically by plugin,
+ * but `StringMessage` and `IntMessage` require usual registration, as described in [PolymorphicSerializer] documentation:
+ *
+ * ```
+ * val abstractContext = SerializersModule {
+ *     polymorphic(ProtocolWithAbstractClass::class, ProtocolWithAbstractClass.Message::class) {
+ *         ProtocolWithAbstractClass.Message.IntMessage::class with ProtocolWithAbstractClass.Message.IntMessage.serializer()
+ *         ProtocolWithAbstractClass.Message.StringMessage::class with ProtocolWithAbstractClass.Message.StringMessage.serializer()
+ *         // no need to register ProtocolWithAbstractClass.ErrorMessage
+ *     }
+ * }
+ * ```
+ *
+ */
+public class SealedClassSerializer<T : Any>(
+    serialName: String,
+    override val baseClass: KClass<T>,
+    subclasses: Array<KClass<out T>>,
+    subSerializers: Array<KSerializer<out T>>
```

Comment:
Yes, it is used directly in the codegen. Unfortunately, it is not possible [for now] to make it non-primary because it would require to change plugin code, too. We anyway have a plan to refactor all such constructors, don't we?

Reply:
Okay, let's then make this constructor internal (via annotation) and provide a secondary constructor from a single map